golang/Chapter06/model/client.go

```diff
@@ -0,0 +1,755 @@
+
+		pipeline.ZAdd("seen:" + recipient, &redis.Z{Member:chatId, Score:seenId})
+		if minId != nil {
+			pipeline.ZRemRangeByScore("msgs:" + chatId, string(0), strconv.Itoa(int(minId[0].Score)))
+		}
+		result[i] = Messages{ChatId:chatId, ChatMessages:messages}
+	}
+
+	return result
+}
+
+func (c *Client) JoinChat(chatId, user string) {
+	messageId, _ := c.Conn.Get("ids" + chatId).Float64()
+
+	pipeline := c.Conn.TxPipeline()
+	pipeline.ZAdd("chat:" + chatId, &redis.Z{Member:user, Score:messageId})
+	pipeline.ZAdd("seen:" + user, &redis.Z{Member:chatId, Score:messageId})
+	if _, err := pipeline.Exec(); err != nil {
+		log.Println("pipeline err in JoinChat: ", err)
+	}
+}
+
+func (c *Client) LeaveChat(chatId, user string) {
+	res := &redis.IntCmd{}
+	pipeline := c.Conn.TxPipeline()
+	pipeline.ZRem("chat:" + chatId, user)
+	pipeline.ZRem("seen:" + user, chatId)
+	res = pipeline.ZCard("chat:" + chatId)
+	if _, err := pipeline.Exec(); err != nil {
+		log.Println("pipeline err in LeaveChat: ", err)
+	}
+
+	if res == nil {
+		pipeline.Del("msgs:" + chatId)
+		pipeline.Del("ids:" + chatId)
+		if _, err := pipeline.Exec(); err != nil {
+			log.Println("pipeline err in LeaveChat: ", err)
+		}
+	} else {
+		oldest := c.Conn.ZRangeWithScores("chat:" + chatId, 0, 0).Val()[0]
+		c.Conn.ZRemRangeByScore("msgs:" + chatId, "0", strconv.Itoa(int(oldest.Score)))
+	}
+}
+
+func (c *Client) CopyLogsToRedis(path, channel string, count int, limit int64, quitWhenDone bool) {
+	var bytesInRedis int64 = 0
+	waiting := []os.FileInfo{}
+	recipies := make([]string, count)
+	for i := 0; i < count; i++ {
+		recipies[i] = strconv.Itoa(i)
+	}
+	c.CreateChat("source", &recipies, "", channel)
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Println("get dir err in CopyLogsToRedis: ", err)
+		return
+	}
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name() < files[j].Name()
+	})
+	for _, logfile := range files {
+		fullPath := filepath.Join(path, logfile.Name())
+
+		finfo, _ := os.Stat(fullPath)
+		fsize := finfo.Size()
+		for bytesInRedis + fsize > limit {
+			cleaned := c.clean(channel, &waiting, count)
+			if cleaned != 0 {
+				bytesInRedis -= cleaned
+			} else {
+				time.Sleep(250 * time.Millisecond)
+			}
+		}
+
+		file, err := os.Open(fullPath)
+		if err != nil {
+			log.Fatalln("open file in CopyLogsToRedis err: ", err)
+			return
+		}
+
+		byteSlice := make([]byte, int(math.Pow(2, 17)))
+		for {
+			temp, _ := file.Read(byteSlice)
+			byteSlice = byteSlice[:temp]
+			if temp == 0 {
+				break
+			}
+			c.Conn.Append(channel + logfile.Name(), string(byteSlice))
+		}
+		c.SendMessage(channel, "source", logfile.Name())
+
+		bytesInRedis += fsize
+		waiting = append(waiting, logfile)
+	}
+
+	if quitWhenDone {
+		c.SendMessage(channel, "source", ":done")
+	}
+
+	for len(waiting) != 0 {
+		cleaned := c.clean(channel, &waiting, count)
+		if cleaned != 0 {
+			bytesInRedis -= cleaned
+		} else {
+			time.Sleep(250 * time.Millisecond)
+		}
+	}
+}
```

Comment:
This logs transfer thing was also one other thing I wish I didn't write. Oof. Yours looks fine :)